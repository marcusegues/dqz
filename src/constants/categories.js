// @flow
import type { Categories } from '../types/basket';

export const CATEGORIES: Categories = {
  MEAT_AND_MEAT_PRODUCTS: 'MEAT_AND_MEAT_PRODUCTS',
  BUTTER_OR_CREAM: 'BUTTER_OR_CREAM',
  OILS_FATS_MARGARINE: 'OILS_FATS_MARGARINE',
  OTHER_FOODS: 'OTHER_FOODS',
  ALCOHOL_BELOW_18: 'ALCOHOL_BELOW_18',
  ALCOHOL_ABOVE_18: 'ALCOHOL_ABOVE_18',
  CIGARETTES_AND_CIGARS: 'CIGARETTES_AND_CIGARS',
  OTHER_TOBACCO: 'OTHER_TOBACCO',
  MEDICINE: 'MEDICINE',
  BOOKS: 'BOOKS',
  MAGAZINES: 'MAGAZINES',
  FLOWERS_AND_PLANTS: 'FLOWERS_AND_PLANTS',
  ANIMAL_FEED: 'ANIMAL_FEED',
  FERTILIZER: 'FERTILIZER',
  OTHER_GOODS: 'OTHER_GOODS',
};
