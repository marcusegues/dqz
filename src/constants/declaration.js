export const MAX_DECLARED_CHF = 20000;
