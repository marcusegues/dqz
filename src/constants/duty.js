export const dutyFreeAllowances = {
  MEAT_AND_MEAT_PRODUCTS: 1,
  BUTTER_OR_CREAM: 1,
  OILS_FATS_MARGARINE: 5,
  ALCOHOL_BELOW_18: 5,
  ALCOHOL_ABOVE_18: 1,
  CIGARETTES_AND_CIGARS: 250,
  OTHER_TOBACCO: 250,
};

export const MEAT_AND_MEAT_PRODUCTS_FIRST_DUTY_THRESHOLD = 10;
export const MEAT_AND_MEAT_PRODUCTS_DUTY_UNDER_THRESHOLD = 17;
export const MEAT_AND_MEAT_PRODUCTS_DUTY_OVER_THRESHOLD = 23;

export const BUTTER_OR_CREAM_DUTY = 16;
export const OILS_FATS_MARGARINE_DUTY = 2;
export const ALCOHOL_BELOW_18_DUTY = 2;
export const ALCOHOL_ABOVE_18_DUTY = 15;
export const CIGARETTES_AND_CIGARS_DUTY = 0.25;
export const OTHER_TOBACCO_DUTY = 0.1;
