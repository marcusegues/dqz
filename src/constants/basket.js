// @flow
import { CATEGORIES_LIST } from './categories';
import type { Basket } from '../types/basket';

let emptyBasket: Basket = {};
CATEGORIES_LIST.forEach(category => {
  emptyBasket[category] = {
    quantity: 0,
    duty: 0,
    values: [],
  };
});

export { emptyBasket };
