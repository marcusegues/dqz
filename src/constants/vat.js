// @flow
import { CATEGORIES } from './categories';
import type { VatRates } from '../types/basket';

// any changes to the vat rates go here
export const VAT_RATES: VatRates = {
  [CATEGORIES.MEAT_AND_MEAT_PRODUCTS]: 0.025,
  [CATEGORIES.BUTTER_OR_CREAM]: 0.025,
  [CATEGORIES.OILS_FATS_MARGARINE]: 0.025,
  [CATEGORIES.OTHER_FOODS]: 0.025,
  [CATEGORIES.ALCOHOL_BELOW_18]: 0.08,
  [CATEGORIES.ALCOHOL_ABOVE_18]: 0.08,
  [CATEGORIES.CIGARETTES_AND_CIGARS]: 0.08,
  [CATEGORIES.OTHER_TOBACCO]: 0.08,
  [CATEGORIES.MEDICINE]: 0.025,
  [CATEGORIES.BOOKS]: 0.025,
  [CATEGORIES.MAGAZINES]: 0.025,
  [CATEGORIES.FLOWERS_AND_PLANTS]: 0.025,
  [CATEGORIES.ANIMAL_FEED]: 0.025,
  [CATEGORIES.FERTILIZER]: 0.025,
  [CATEGORIES.OTHER_GOODS]: 0.08,
};
