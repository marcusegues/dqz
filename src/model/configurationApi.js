// @flow

import Immutable from 'immutable';
import { v4 } from 'uuid';
import type { List as ImmutableListType } from 'immutable';
import type {
  Amounts,
  AmountWithId,
  Basket,
  Category,
  CategoryBasketItem,
  People,
  Quantity,
} from './types/basketPeopleAmountsTypes';
import { categoriesArray } from './constants';
import {
  makeAmountsOfCurrencyRecord,
  makeCategoryBasketItemRecord,
  makePeopleRecord,
} from './types/basketPeopleAmountsTypes';
import type { Currency } from './currencies';
import { flatAllAmounts, getConvertedLocalTimeToSwiss } from './utils';
import type { FlatAmount } from './utils';

const emptyItem: CategoryBasketItem = makeCategoryBasketItemRecord();

/**
 * Empty Basket
 * @type Basket
 */
export const emptyBasket: Basket = Immutable.Map(
  categoriesArray.map(c => [c, emptyItem])
);

// QUANTITIES
/**
 * Adds a quantity in a basket for a given category
 * @param basket
 * @param category
 * @param quantity
 * @returns Basket
 */
export const addQuantity = (
  basket: Basket,
  category: Category,
  quantity: number
): Basket => {
  if (quantity <= 0) {
    return basket;
  }
  return basket.updateIn([category, 'volume', 'quantities'], q =>
    q.push({
      number: quantity,
      date: getConvertedLocalTimeToSwiss().toString(),
    })
  );
};

/**
 * Deletes a quantity in a basket for a given category at index
 * Note: if index > size of quantities list, quantities won't change.
 * This exception should be handled elsewhere.
 * @param basket
 * @param category
 * @param index
 * @returns Basket
 */
export const deleteQuantity = (
  basket: Basket,
  category: Category,
  index: number
): Basket =>
  basket.updateIn([category, 'volume', 'quantities'], q => q.delete(index));

/**
 * Reset all quantites in a basket for a given category.
 * @param basket
 * @param category
 * @returns Basket
 */
export const resetQuantities = (basket: Basket, category: Category): Basket =>
  basket.setIn([category, 'volume', 'quantities'], Immutable.List());

/**
 * Reset all quantites in a basket for a given category.
 * @param basket
 * @param categories (array of category)
 * @returns Basket
 */
export const resetQuantitiesMultipleCategories = (
  basket: Basket,
  categories: Array<Category>
): Basket =>
  basket.withMutations(bskt => {
    categories.forEach(category => {
      bskt.setIn([category, 'volume', 'quantities'], Immutable.List());
    });
    return bskt;
  });

/**
 * Helper method, not to be used in the App.
 * Sets several quantites for the given basket in different categories
 * @param basket
 * @param pairs
 * @returns Basket
 */
export const addQuantities = (
  basket: Basket,
  pairs: Array<{ category: Category, quantity: number }>
): Basket =>
  basket.withMutations(bskt => {
    pairs.forEach(p => addQuantity(bskt, p.category, p.quantity));
  });

/**
 * Gets all quantities for the basket's category
 * @param basket
 * @param category
 * @returns List of quantities (object)
 */
export const getQuantities = (
  basket: Basket,
  category: Category
): ImmutableListType<Quantity> =>
  basket.getIn([category, 'volume', 'quantities'], Immutable.List());

/**
 * Helper to get the quantity number from a quantity
 * @param quantity
 * @returns {number}
 */
export const getQuantityNumber = (quantity: Quantity): number =>
  quantity.number || 0;

/**
 * Helper to get the quantity date (string) from a quantity
 * @param quantity
 * @returns {*}
 */
export const getQuantityDate = (quantity: Quantity): string => quantity.date;

/**
 * Gets the total quantity for the basket's category
 * @param basket
 * @param category
 * @returns number
 */
export const getTotalQuantity = (
  basket: Basket,
  category: Category
): number => basket
    .getIn([category, 'volume', 'quantities'], Immutable.List())
    .reduce((a, v) => a + getQuantityNumber(v), 0);

// AMOUNTS
export const initAmounts: Amounts = Immutable.Map();

/**
 * Helper to create currency section if not present
 */
const addCurrencyIfNeeded = (amounts: Amounts, currency: Currency): Amounts => {
  if (amounts.has(currency)) {
    return amounts;
  }
  return amounts.set(currency, makeAmountsOfCurrencyRecord());
};

/**
 * Adds an amount in a given category of a basket
 * @param amounts
 * @param currency
 * @param amount
 * @returns Amounts type
 */
export const addAmount = (
  amounts: Amounts,
  currency: Currency,
  amount: number
): Amounts =>
  addCurrencyIfNeeded(amounts, currency).updateIn(
    [currency, 'amounts'],
    Immutable.List(),
    list => list.push({ amount, id: v4() })
  );

/**
 * Resets the amounts in a basket's category
 * @param amounts
 * @param currency
 * @returns Amounts type
 */
export const resetAmounts = (amounts: Amounts, currency: Currency): Amounts =>
  addCurrencyIfNeeded(amounts, currency).setIn(
    [currency, 'amounts'],
    Immutable.List()
  );

/**
 * Deletes amount with a given ID
 * @param amounts
 * @param id
 * @returns {Amounts}
 */
export const deleteAmount = (amounts: Amounts, id: string): Amounts => {
  const flat = flatAllAmounts(amounts);
  const element: ?FlatAmount = flat.find(a => a.id === id);
  if (!element) {
    return amounts;
  }
  const amountsKey = element.large ? 'largeAmounts' : 'amounts';
  const amts = amounts.getIn([element.currency, amountsKey], Immutable.List());
  const listIdx = amts.findIndex(v => v.id === id);
  const newAmounts = amts.delete(listIdx);
  return amounts.setIn([element.currency, amountsKey], newAmounts);
};

/**
 * Gets amount at index
 * @param amounts
 * @param id
 * @returns The amount (type FlatAmount) or undefined
 */
export const getAmount = (amounts: Amounts, id: string): FlatAmount => {
  const flat = flatAllAmounts(amounts);
  const el = flat.find(a => a.id === id);
  if (el) {
    return el;
  }
  return { amount: 0, currency: 'CHF', large: false, id: 'N/A' };
};

/**
 * Gets the amounts of a basket's category
 * @param amounts
 * @param currency
 * @returns ImmutableListType<number>
 */
export const getAmounts = (
  amounts: Amounts,
  currency: Currency
): ImmutableListType<AmountWithId> =>
  amounts.getIn([currency, 'amounts'], Immutable.List());

// AMOUNTS LARGE
/**
 * Adds a large amount in a given category of a basket
 * @param amounts
 * @param currency
 * @param amountLarge
 * @returns Amount type
 */
export const addLargeAmount = (
  amounts: Amounts,
  currency: Currency,
  amountLarge: number
): Amounts =>
  addCurrencyIfNeeded(amounts, currency).updateIn(
    [currency, 'largeAmounts'],
    Immutable.List(),
    list => list.push({ amount: amountLarge, id: v4() })
  );

/**
 * Resets the large amounts in a basket's category
 * @param amounts
 * @param currency
 * @returns Basket
 */
export const resetLargeAmounts = (
  amounts: Amounts,
  currency: Currency
): Amounts =>
  addCurrencyIfNeeded(amounts, currency).setIn(
    [currency, 'largeAmounts'],
    Immutable.List()
  );

/**
 * Gets the large amounts of a basket's category
 * @param amounts
 * @param currency
 * @returns ImmutableListType<AmountWithId>
 */
export const getLargeAmounts = (
  amounts: Amounts,
  currency: Currency
): ImmutableListType<AmountWithId> =>
  amounts.getIn([currency, 'largeAmounts'], Immutable.List());

/**
 * Returns an init people configuration (1 adult, no minor)
 */
export const initPeople: People = makePeopleRecord({
  adults: 1,
  minors: 0,
});

/**
 * Gets the people in the party (adult + minor)
 * @param party
 * @returns number
 */
export const getTotalPeople = (party: People): number =>
  party.get('adults') + party.get('minors');

/**
 * Gets the adult people in the party
 * @param party
 * @returns number
 */
export const getAdultPeople = (party: People): number => party.get('adults');

/**
 * Gets the minor people in the party
 * @param party
 * @returns number
 */
export const getMinorPeople = (party: People): number => party.get('minors');

/**
 * Adds an adult to the party
 * @param party
 * @returns People Type
 */
export const addAdult = (party: People): People =>
  party.update('adults', a => a + 1);

/**
 * Sets the number of adults in a party (non-neg)
 * @param party
 * @param n
 * @returns People Type
 */
export const setAdultPeople = (party: People, n: number): People =>
  party.set('adults', Math.max(0, n));

/**
 * Subtracts an adult from the party (non-negative result)
 * @param party
 * @returns People Type
 */
export const subtractAdult = (party: People): People =>
  party.update('adults', a => Math.max(a - 1, 0));

/**
 * Adds a minor to the party
 * @param party
 * @returns People Type
 */
export const addMinor = (party: People): People =>
  party.update('minors', m => m + 1);

/**
 * Sets the number of adults in a party (non-neg)
 * @param party
 * @param n
 * @returns People Type
 */
export const setMinorPeople = (party: People, n: number): People =>
  party.set('minors', Math.max(0, n));

/**
 * Subtracts a minor from the party (non-negative result)
 * @param party
 * @returns People Type
 */
export const subtractMinor = (party: People): People =>
  party.update('minors', m => Math.max(m - 1, 0));

/**
 * Set all people directly (non-negative)
 * @param party
 * @param adults
 * @param minors
 * @returns People Type
 */
export const setPeople = (
  party: People,
  adults: number,
  minors: number
): People => {
  const p1 = party.set('minors', Math.max(0, minors));
  return p1.set('adults', Math.max(0, adults));
};
