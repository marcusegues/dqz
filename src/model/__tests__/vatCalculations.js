// @flow
import {
  calculateAllowancesExceptLast,
  calculateVat,
  calculateVatLargeItems,
  calculateVatNormalItems,
  applyAllowancesNotLastPerson,
  summarizeByVatBracket,
  vatByCategory,
  totalAmounts,
  applyAllowancesLastPerson,
} from '../vatCalculations';
import {
  ezvBasket1,
  ezvBasket10,
  ezvBasket11,
  ezvBasket12,
  ezvBasket13,
  ezvBasket2,
  ezvBasket3,
  ezvBasket4,
  ezvBasket5,
  ezvBasket6,
  ezvBasket7,
  ezvBasket8,
  ezvBasket9,
  presentationBasket1,
  presentationBasket2,
  sampleBasket1,
  sampleBasket2,
  sampleBasket3,
  sampleBasket4,
} from './fullBaskets';
import { addAdult, addMinor, initPeople } from '../configurationApi';
import { IndividualAllowance } from '../constants';

const b1 = summarizeByVatBracket(sampleBasket1);
const b2 = summarizeByVatBracket(sampleBasket2);
const twoAdults = addAdult(initPeople);
const threeAdults = addAdult(twoAdults);

describe('VAT Calculations', () => {
  test('it sums vat for all vat-categories', () => {});

  test('it sums all the categories as expected', () => {
    expect(b1).toMatchSnapshot();
  });

  test('it calculates vat by category', () => {
    expect(vatByCategory(sampleBasket1)).toMatchSnapshot();
    expect(vatByCategory(sampleBasket2)).toMatchSnapshot();
  });

  test('it calculates the vat of large items (no deduction)', () => {
    expect(calculateVatLargeItems(b1)).toBeCloseTo(457.55);
  });

  test('it calculates the vat of normal items (no deduction)', () => {
    expect(calculateVatNormalItems(b1)).toBe(8);
  });

  test('it sums all the categories and subtracts allowances (pre-last) as expected', () => {
    expect(applyAllowancesNotLastPerson(b1, twoAdults)).toMatchSnapshot();
  });

  test('it sums all the categories and subtracts allowances (pre-last) as expected', () => {
    expect(applyAllowancesNotLastPerson(b2, twoAdults)).toMatchSnapshot();
  });

  test('it sums all the categories and subtracts allowances (last) as expected', () => {
    expect(
      applyAllowancesLastPerson(b1, totalAmounts(sampleBasket1))
    ).toMatchSnapshot();
  });

  test('it sums all the categories and subtracts allowances (last) as expected', () => {
    expect(
      applyAllowancesLastPerson(b2, totalAmounts(sampleBasket2))
    ).toMatchSnapshot();
  });

  test('calculate allowance correctly', () => {
    expect(calculateAllowancesExceptLast(initPeople)).toBe(0);
    expect(calculateAllowancesExceptLast(addAdult(initPeople))).toBe(
      IndividualAllowance
    );
    expect(calculateAllowancesExceptLast(addMinor(initPeople))).toBe(
      IndividualAllowance
    );
  });

  test('it creates a correct vat summary', () => {
    expect(calculateVat(sampleBasket1, initPeople)).toMatchSnapshot();
    expect(calculateVat(sampleBasket2, initPeople)).toMatchSnapshot();
    expect(calculateVat(sampleBasket3, initPeople).get('totalVat')).toBeCloseTo(
      53.45
    );
    expect(calculateVat(sampleBasket4, initPeople).get('totalVat')).toBeCloseTo(
      35.9975
    );
  });

  test('it works with EZV examples', () => {
    expect(calculateVat(ezvBasket1, initPeople)).toMatchSnapshot();
    expect(calculateVat(ezvBasket2, initPeople)).toMatchSnapshot();
    expect(calculateVat(ezvBasket3, twoAdults)).toMatchSnapshot();
    expect(calculateVat(ezvBasket4, twoAdults)).toMatchSnapshot();
    expect(calculateVat(ezvBasket5, twoAdults)).toMatchSnapshot();
    expect(calculateVat(ezvBasket6, twoAdults)).toMatchSnapshot();
    expect(calculateVat(ezvBasket7, twoAdults)).toMatchSnapshot();
    expect(calculateVat(ezvBasket8, twoAdults)).toMatchSnapshot();
    expect(calculateVat(ezvBasket9, twoAdults)).toMatchSnapshot();
    expect(calculateVat(ezvBasket10, threeAdults)).toMatchSnapshot();
    expect(calculateVat(ezvBasket11, threeAdults)).toMatchSnapshot();
    expect(calculateVat(ezvBasket12, threeAdults)).toMatchSnapshot();
    expect(calculateVat(ezvBasket13, threeAdults)).toMatchSnapshot();
    expect(calculateVat(presentationBasket1, initPeople)).toMatchSnapshot();
    expect(
      calculateVat(presentationBasket2, addMinor(initPeople))
    ).toMatchSnapshot();
  });
});
