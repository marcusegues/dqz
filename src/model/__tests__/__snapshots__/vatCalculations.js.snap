// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VAT Calculations it calculates vat by category 1`] = `
Immutable.Map {
  "Butter": 16.558500000000002,
  "Oils": 0,
  "Cigarettes": 0,
  "AlcSoft": 0,
  "Fertilizer": 0,
  "AlcHard": 0,
  "Meds": 0,
  "Books": 0,
  "Meat": 1.075,
  "OtherFood": 0,
  "Other": 0,
  "Tobacco": 447.92,
  "Flowers": 0,
  "AnimalFeed": 0,
  "Magazines": 0,
}
`;

exports[`VAT Calculations it calculates vat by category 2`] = `
Immutable.Map {
  "Butter": 38.43350000000001,
  "Oils": 0,
  "Cigarettes": 0,
  "AlcSoft": 0,
  "Fertilizer": 0,
  "AlcHard": 0,
  "Meds": 0,
  "Books": 0,
  "Meat": 1.075,
  "OtherFood": 0,
  "Other": 0,
  "Tobacco": 447.92,
  "Flowers": 0,
  "AnimalFeed": 0,
  "Magazines": 0,
}
`;

exports[`VAT Calculations it creates a correct vat summary 1`] = `
Immutable.Record {
  "totalVat": 465.55350000000004,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 16.558500000000002,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 1.075,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 447.92,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it creates a correct vat summary 2`] = `
Immutable.Record {
  "totalVat": 487.42850000000004,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 38.43350000000001,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 1.075,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 447.92,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it sums all the categories and subtracts allowances (last) as expected 1`] = `
Immutable.Record {
  "large": Immutable.OrderedMap {
    0.08: Immutable.List [
      5555,
    ],
    0.025: Immutable.List [
      525,
    ],
  },
  "normal": Immutable.OrderedMap {
    0.08: Immutable.List [
      11,
      33,
    ],
    0.025: Immutable.List [
      12,
      31,
      12.34,
      125,
    ],
  },
}
`;

exports[`VAT Calculations it sums all the categories and subtracts allowances (last) as expected 2`] = `
Immutable.Record {
  "large": Immutable.OrderedMap {
    0.08: Immutable.List [
      5555,
    ],
    0.025: Immutable.List [
      525,
    ],
  },
  "normal": Immutable.OrderedMap {
    0.08: Immutable.List [
      11,
      33,
    ],
    0.025: Immutable.List [
      12,
      31,
      12.34,
      125,
      125,
      125,
      125,
      125,
      125,
      125,
      125,
    ],
  },
}
`;

exports[`VAT Calculations it sums all the categories and subtracts allowances (pre-last) as expected 1`] = `
Immutable.Record {
  "large": Immutable.OrderedMap {
    0.08: Immutable.List [
      5555,
    ],
    0.025: Immutable.List [
      525,
    ],
  },
  "normal": Immutable.OrderedMap {
    0.08: Immutable.List [
      0,
      0,
    ],
    0.025: Immutable.List [
      0,
      0,
      0,
      0,
    ],
  },
}
`;

exports[`VAT Calculations it sums all the categories and subtracts allowances (pre-last) as expected 2`] = `
Immutable.Record {
  "large": Immutable.OrderedMap {
    0.08: Immutable.List [
      5555,
    ],
    0.025: Immutable.List [
      525,
    ],
  },
  "normal": Immutable.OrderedMap {
    0.08: Immutable.List [
      0,
      0,
    ],
    0.025: Immutable.List [
      0,
      0,
      0,
      0,
      49.34,
      125,
      125,
      125,
      125,
      125,
      125,
    ],
  },
}
`;

exports[`VAT Calculations it sums all the categories as expected 1`] = `
Immutable.Record {
  "large": Immutable.OrderedMap {
    0.08: Immutable.List [
      5555,
    ],
    0.025: Immutable.List [
      525,
    ],
  },
  "normal": Immutable.OrderedMap {
    0.08: Immutable.List [
      11,
      33,
    ],
    0.025: Immutable.List [
      12,
      31,
      12.34,
      125,
    ],
  },
}
`;

exports[`VAT Calculations it works with EZV examples 1`] = `
Immutable.Record {
  "totalVat": 0,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 7.5,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 2`] = `
Immutable.Record {
  "totalVat": 10,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 10,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 3`] = `
Immutable.Record {
  "totalVat": 0,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 15,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 4`] = `
Immutable.Record {
  "totalVat": 15,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 22.5,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 5`] = `
Immutable.Record {
  "totalVat": 10,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 10,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 6`] = `
Immutable.Record {
  "totalVat": 8,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 8,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 7`] = `
Immutable.Record {
  "totalVat": 0,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 11,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 8`] = `
Immutable.Record {
  "totalVat": 9,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 16.5,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 9`] = `
Immutable.Record {
  "totalVat": 17.5,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 25,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 10`] = `
Immutable.Record {
  "totalVat": 20,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 20,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 11`] = `
Immutable.Record {
  "totalVat": 10,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 25,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 12`] = `
Immutable.Record {
  "totalVat": 12.5,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 22.5,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 13`] = `
Immutable.Record {
  "totalVat": 22.5,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 37.5,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 14`] = `
Immutable.Record {
  "totalVat": 53.431,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0.595,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 7.616,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 7.14,
    "OtherFood": 0,
    "Other": 38.08,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`VAT Calculations it works with EZV examples 15`] = `
Immutable.Record {
  "totalVat": 12.135,
  "vatByCategoryRaw": Immutable.Map {
    "Butter": 0.595,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 4.76,
    "Meds": 0,
    "Books": 0,
    "Meat": 1.19,
    "OtherFood": 10.412500000000001,
    "Other": 19.04,
    "Tobacco": 0,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;
